File: test_telegram_replies.py
#!/usr/bin/env python3
import asyncio
import json
from pyrogram import Client, filters
from pyrogram.types import Message
import os
from dotenv import load_dotenv

load_dotenv()

# Get credentials from environment
api_id = os.getenv('TELEGRAM_API_ID')
api_hash = os.getenv('TELEGRAM_API_HASH')
group_id = int(os.getenv('TELEGRAM_GROUP_ID', '0'))

if not api_id or not api_hash:
    print("Missing TELEGRAM_API_ID or TELEGRAM_API_HASH in .env")
    exit(1)

# Create client
app = Client(
    "reply_test_bot",
    api_id=api_id,
    api_hash=api_hash
)

# Track messages
message_map = {}

@app.on_message(filters.chat(group_id))
async def handle_message(client: Client, message: Message):
    print("\n=== NEW MESSAGE (Pyrogram) ===")
    print(f"Message ID: {message.id}")
    print(f"From: {message.from_user.username if message.from_user else 'Unknown'}")
    print(f"Text: {message.text}")
    print(f"Topic ID: {message.topic.id if message.topic else 'none'}")
    print(f"Has reply: {message.reply_to_message is not None}")
    print(f"Reply to message ID: {message.reply_to_message_id}")
    print(f"Reply to top message ID: {message.reply_to_top_message_id}")
    
    if message.reply_to_message:
        print("\n--- REPLY DETAILS (Pyrogram) ---")
        reply = message.reply_to_message
        print(f"Reply message ID: {reply.id}")
        print(f"Reply text: {reply.text}")
        print(f"Reply from: {reply.from_user.username if reply.from_user else 'Unknown'}")
        print(f"Reply is bot: {reply.from_user.is_bot if reply.from_user else 'Unknown'}")
        
        # Check if it's a topic reply
        if message.topic and reply.id == message.topic.id:
            print("This is a reply to the topic itself, not a real message reply")
    
    # Store message
    message_map[message.id] = message
    
    # Test command
    if message.text == "/pyrogram_test":
        test_msg = await message.reply("Reply to this message to test Pyrogram reply detection!")
        print(f"\nSent test message {test_msg.id}")
        message_map[test_msg.id] = test_msg
    
    # Debug dump
    if message.text and "debug" in message.text:
        print("\n=== FULL MESSAGE DUMP (Pyrogram) ===")
        # Convert to dict for JSON serialization
        msg_dict = {
            "id": message.id,
            "text": message.text,
            "date": str(message.date),
            "chat_id": message.chat.id,
            "from_user": {
                "id": message.from_user.id,
                "username": message.from_user.username,
                "is_bot": message.from_user.is_bot
            } if message.from_user else None,
            "reply_to_message_id": message.reply_to_message_id,
            "reply_to_top_message_id": message.reply_to_top_message_id,
            "topic": {
                "id": message.topic.id,
                "title": message.topic.title
            } if message.topic else None,
            "message_thread_id": message.message_thread_id
        }
        print(json.dumps(msg_dict, indent=2))

async def main():
    print("Starting Pyrogram reply test...")
    print(f"Group ID: {group_id}")
    
    async with app:
        # Get chat info
        try:
            chat = await app.get_chat(group_id)
            print(f"Chat info: {chat.title} (type: {chat.type})")
        except Exception as e:
            print(f"Failed to get chat info: {e}")
        
        print("\nBot is running. Send messages in the Telegram group to test.")
        print("Commands:")
        print("  /pyrogram_test - Send a test message you can reply to")
        print("  Include 'debug' in your message for full message dump")
        print("\nPress Ctrl+C to stop.")
        
        # Keep the script running
        await app.idle()

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\nStopping bot...")

File: test_telegram_replies.js
const TelegramBot = require('node-telegram-bot-api');
require('dotenv').config();

const botToken = process.env.TELEGRAM_BOT_TOKEN;
const groupId = process.env.TELEGRAM_GROUP_ID;

if (!botToken || !groupId) {
  console.error('Missing TELEGRAM_BOT_TOKEN or TELEGRAM_GROUP_ID in .env');
  process.exit(1);
}

const bot = new TelegramBot(botToken, { polling: false });

console.log('Starting Telegram reply test...');
console.log('Bot token:', botToken.substring(0, 10) + '...');
console.log('Group ID:', groupId);

// Start polling with enhanced settings
bot.startPolling({
  polling: {
    interval: 1000,
    autoStart: true,
    params: {
      timeout: 25,
      allowed_updates: ['message', 'edited_message']
    }
  }
}).then(() => {
  console.log('Bot started polling successfully');
}).catch(err => {
  console.error('Failed to start polling:', err);
});

// Get bot info
bot.getMe().then(me => {
  console.log('Bot info:', {
    id: me.id,
    username: me.username,
    first_name: me.first_name
  });
}).catch(err => {
  console.error('Failed to get bot info:', err);
});

// Get chat info
bot.getChat(groupId).then(chat => {
  console.log('Chat info:', {
    id: chat.id,
    title: chat.title,
    type: chat.type,
    is_forum: chat.is_forum
  });
}).catch(err => {
  console.error('Failed to get chat info:', err);
});

// Track messages for reply testing
const messageMap = new Map();

bot.on('message', (msg) => {
  if (msg.chat.id.toString() !== groupId) return;
  
  console.log('\n=== NEW MESSAGE ===');
  console.log('Message ID:', msg.message_id);
  console.log('From:', msg.from?.username || msg.from?.first_name || 'Unknown');
  console.log('Text:', msg.text);
  console.log('Thread ID:', msg.message_thread_id || 'none');
  console.log('Has reply_to_message:', !!msg.reply_to_message);
  
  if (msg.reply_to_message) {
    console.log('\n--- REPLY DETAILS ---');
    console.log('Reply to message ID:', msg.reply_to_message.message_id);
    console.log('Reply to text:', msg.reply_to_message.text);
    console.log('Reply to from:', msg.reply_to_message.from?.username || msg.reply_to_message.from?.first_name);
    console.log('Reply is bot message:', msg.reply_to_message.from?.is_bot);
    console.log('Is reply to thread itself:', msg.reply_to_message.message_id === msg.message_thread_id);
    
    // Full structure dump
    console.log('\nFull reply_to_message structure:');
    console.log(JSON.stringify(msg.reply_to_message, null, 2));
  }
  
  // Store message for reference
  messageMap.set(msg.message_id, msg);
  
  // Test command to send a message that can be replied to
  if (msg.text === '/testreply') {
    const options = {};
    if (msg.message_thread_id) {
      options.message_thread_id = msg.message_thread_id;
    }
    
    bot.sendMessage(msg.chat.id, 'Reply to this message to test reply detection!', options)
      .then(sentMsg => {
        console.log(`\nSent test message ${sentMsg.message_id} in thread ${sentMsg.message_thread_id || 'general'}`);
        messageMap.set(sentMsg.message_id, sentMsg);
      })
      .catch(err => {
        console.error('Failed to send test message:', err);
      });
  }
  
  // Full message dump for debugging
  if (msg.text && msg.text.includes('debug')) {
    console.log('\n=== FULL MESSAGE DUMP ===');
    console.log(JSON.stringify(msg, null, 2));
  }
});

bot.on('polling_error', (error) => {
  console.error('Polling error:', error);
});

bot.on('error', (error) => {
  console.error('Bot error:', error);
});

console.log('\nBot is running. Send messages in the Telegram group to test.');
console.log('Commands:');
console.log('  /testreply - Send a test message you can reply to');
console.log('  Include "debug" in your message for full message dump');
console.log('\nPress Ctrl+C to stop.');

// Graceful shutdown
process.on('SIGINT', () => {
  console.log('\nStopping bot...');
  bot.stopPolling();
  process.exit(0);
});